### PR TITLE
[RISCV][ISel] Use vaaddu with rounding mode rdn for ISD::AVGFLOORU.

### PR DESCRIPTION
**Pull request by sun-jacobi**
This patch aims to use `vaaddu` with rounding mode rdn (i.e `vxrm[1:0] = 0b10`) for `ISD::AVGFLOORU`.

### Source code 
```
define <8 x i8> @vaaddu_auto(ptr %x, ptr %y, ptr %z) {
  %xv = load <8 x i8>, ptr %x, align 2
  %yv = load <8 x i8>, ptr %y, align 2
  %xzv = zext <8 x i8> %xv to <8 x i16>
  %yzv = zext <8 x i8> %yv to <8 x i16>
  %add = add nuw nsw <8 x i16> %xzv, %yzv
  %div = lshr <8 x i16> %add, <i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1, i16 1>
  %ret = trunc <8 x i16> %div to <8 x i8>
  ret <8 x i8> %ret 
}
```

### Before this patch 
```
vaaddu_auto: 
        vsetivli        zero, 8, e8, mf2, ta, ma
        vle8.v  v8, (a0)
        vle8.v  v9, (a1)
        vwaddu.vv       v10, v8, v9
        vnsrl.wi        v8, v10, 1
        ret
```
### After this patch 
```
vaaddu_auto: 
	vsetivli	zero, 8, e8, mf2, ta, ma
	vle8.v	v8, (a0)
	vle8.v	v9, (a1)
	csrwi	vxrm, 2
	vaaddu.vv	v8, v8, v9
	ret
```

### Note on signed averaging addition

Based on the rvv spec,  there is also a variant for signed averaging addition called `vaadd`.   
But AFAIU, no matter in which rounding mode, we cannot achieve the semantic of signed averaging addition through `vaadd`.   
Thus this patch only introduces `vaaddu`

### CODE REVIEW

**Comment by topperc** [comment]
I don't think this is the right way to do this. We should make ISD::AVGFLOORU Custom for RISC-V and let DAGCombiner create an ISD::AVGFLOORU node that we can custom lower to a new RISCVISD::VAADDU_VL opcode.

**Comment by sun-jacobi** [comment]
> I don't think this is the right way to do this. We should make ISD::AVGFLOORU Custom for RISC-V and let DAGCombiner create an ISD::AVGFLOORU node that we can custom lower to a new RISCVISD::VAADDU_VL opcode.

Thank you for the advice, I will try this way.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:12440]: Do we not already get ISD::AVGFLOORU nodes from `combineShiftToAVG` in TargetLowering.cpp?

File: llvm/test/CodeGen/RISCV/rvv/vaadd-autogen.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/vaadd-autogen.ll:5]: Nit, I think you can just simplify the invocations to

```suggestion
; RUN: llc -mtriple=riscv32 -mattr=+v -verify-machineinstrs < %s | FileCheck %s
; RUN: llc -mtriple=riscv64 -mattr=+v -verify-machineinstrs < %s | FileCheck %s
```

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:1383]: We don't need this since we don't have the combine anymore

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:1390]: We can restore this too

File: llvm/test/CodeGen/RISCV/rvv/vaaddu-autogen.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/vaaddu-autogen.ll:1]: Could we rename this to `vaaddu-sdnode.ll` to match the other tests? I believe we usually use `vfoo.ll` for intrinsic tests and `vfoo-sdnode.ll` for tests on generic SD nodes. 

We also typically just use scalable vectors since they're the "native" vector type. Fixed vector tests then go into `fixed-vectors-vfoo.ll`.

File: llvm/test/CodeGen/RISCV/rvv/vaaddu-sdnode.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/vaaddu-sdnode.ll:1]: Can we add some tests for vaaddu.vx? Something like

```llvm
define <vscale x 8 x i8> @vaaddu_vx_nxv8i8(<vscale x 8 x i8> %x, i8 %y) {
  %xzv = zext <vscale x 8 x i8> %x to <vscale x 8 x i16>
  %yhead = insertelement <vscale x 8 x i8> poison, i8 %y, i32 0
  %ysplat = shufflevector <vscale x 8 x i8> %yhead, <vscale x 8 x i8> poison, <vscale x 8 x i32> zeroinitializer
  %yzv = zext <vscale x 8 x i8> %ysplat to <vscale x 8 x i16>
  ...
}
```

File: llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:127]: I think most of the _vl nodes try to correspond to an SDNode in ISDOpcodes.h. Would it make more sense to instead:

- Mark `ISD::AVGFLOORU` as legal for scalable vector types, and custom for fixed length vector types
- Add a `RISCVISD::AVGFLOORU_VL` node with no rounding mode operand, i.e. it uses`SDT_RISCVIntBinOp_VL`
- Map `ISD::AVGFLOORU` to `RISCVISD::AVGFLOORU_VL` in `getRISCVVLOp`
- In `RISCVTargetLowering::LowerOperation`, add a case for `ISD::AVGFLOORU` beside `ISD::SADDSAT` and friends so fixed length vectors gets lowered via `lowerToScalableOp`
- Add a pattern for `avgflooru` in RISCVInstrInfoVSDPatterns.td, and `riscv_avgflooru_vl` in RISCVInstrInfoVVLPatterns.td, where both select `0b10` for the rounding mode in the pattern output

That way we wouldn't need the custom lowering in `RISCVTargetLowering::lowerUnsignedAvgFloor`, and it would follow the same codegen path as the saturating add and subtract instructions.

Comment (sun-jacobi) [reply llvm/lib/Target/RISCV/RISCVInstrInfoVVLPatterns.td:127]: Thank for your detailed advice. I will try this way.

**Review by lukel97 — APPROVED** [review]
Just a couple of nits but otherwise LGTM, thanks! Should probably wait for a second approval too in case anyone has other opinions

**Review by wangpc-pp — APPROVED** [review]
LGTM.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**
